drools-wb-screens/drools-wb-guided-rule-editor/drools-wb-guided-rule-editor-client/src/test/java/org/drools/workbench/screens/guided/rule/client/editor/AttributeSelectorPopupTest.java

```diff
@@ -0,0 +1,153 @@
+@RunWith(GwtMockitoTestRunner.class)
+public class AttributeSelectorPopupTest {
+
+    static String[] attributes = new String[] {"attr1", "attr2", "attr3", "attr4"};
+    static String[] duplicates = new String[] {"attr2", "attr3"};
+    static String metadataName = "mockMetadata";
+
+    @GwtMock
+    TextBox boxMock;
+    @GwtMock
+    ListBox listMock;
+
+    List<String> listMockAttributes;
+
+    MockAttributeSelectorPopup popup;
+
+    @Before
+    public void setUp() {
+        listMockAttributes = new ArrayList<>();
+        popup = spy(new MockAttributeSelectorPopup());
+        when(boxMock.getElement()).thenReturn(mock(Element.class));
+        when(boxMock.getText()).thenReturn(metadataName);
+        when(listMock.getItemCount()).thenAnswer(
+            (InvocationOnMock invocation) -> listMockAttributes.size()
+        );
+        when(listMock.getItemText(anyInt())).thenAnswer(
+            (InvocationOnMock invocation) -> listMockAttributes.get((int) invocation.getArguments()[0])
+        );
+        doAnswer((Answer<Void>)
+            (InvocationOnMock invocation) -> {
+                listMockAttributes.add((String) invocation.getArguments()[0]);
+                return null;
+            }
+        ).when(listMock).addItem(anyString());
+        doAnswer((Answer<Void>)
+            (InvocationOnMock invocation) -> {
+                listMockAttributes.remove((int) invocation.getArguments()[0]);
+                return null;
+            }
+        ).when(listMock).removeItem(anyInt());
+        popup.initialize(boxMock, listMock);
+    }
+
+    @Test
+    public void alreadyUsedAttributesRemoved() {
+        verify(popup).getAttributes();
+        verify(popup).getDuplicates();
+        verify(popup.list, times(4)).addItem(anyString());
+        verify(popup.list, times(2)).removeItem(1);
+    }
+
+    @Test
+    public void alreadyUsedMetadataNotAllowed() {
+        popup.getMetadataHandler().onClick(null);
+        verify(popup).isMetadataUnique(metadataName);
+        verify(popup).metadataNotUniqueMessage(metadataName);
+        verify(popup, never()).handleMetadataAddition(anyString());
+    }
+
+    @Test
+    public void emptyMetadataNotAllowed() {
+        when(boxMock.getText()).thenReturn("");
+        popup.getMetadataHandler().onClick(null);
+        verify(popup, never()).handleMetadataAddition(anyString());
+    }
+
+    @Test
+    public void whitespaceMetadataNotAllowed() {
+        when(boxMock.getText()).thenReturn(" \n\t ");
+        popup.getMetadataHandler().onClick(null);
+        verify(popup, never()).handleMetadataAddition(anyString());
+    }
+
+    private static class MockAttributeSelectorPopup extends AttributeSelectorPopup {
+
+        @Override
+        protected String[] getAttributes() {
+            return attributes;
+        }
+
+        @Override
+        protected String[] getDuplicates() {
+            System.out.println("getting duplicates");
```

Comment:
Sorry, @zkrejcov I missed this the first time round.. it'd be best to remove this type of logging (which I assume you added to help debug your tests, or the like)